### PR TITLE
Fix Deprecated ClassMatcher #11767

### LINKED ISSUE #11767: Backward compat API WebAppContext.getServerClassMatcher() is not updating the WebAppClassloader
**Jetty version(s)**
12.0.9

**Jetty Environment**
ee10 (but could be more)

**Java version/vendor** `(use: java -version)`
Any

**OS type/version**
Any

**Description**
If a project is using the old API `WebAppContext.getServerClassMatcher()` and then modifies the server classes, those changes are not applied to the WebAppClassloader of the WebAppContext.

This applies to Embedded usage and XML deployables.

If the deployable has, for example ...

``` xml
<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">

<Configure class="org.eclipse.jetty.ee10.webapp.WebAppContext">
  <Set name="contextPath">/demo</Set>
  <Set name="war"><Property name="jetty.webapps" />/demo.war
  </Set>
  <Ref refid="Server">
    <Call id="lowResourceMonitor" name="getBean">
      <Arg>
        <Call class="java.lang.Class" name="forName">
          <Arg>org.eclipse.jetty.server.LowResourceMonitor</Arg>
        </Call>
      </Arg>
    </Call>
  </Ref>
  <Call name="setAttribute">
    <Arg>org.eclipse.jetty.server.LowResourceMonitor</Arg>
    <Arg>
      <Ref refid="lowResourceMonitor" />
    </Arg>
  </Call>
  <!-- tweak what classes from the Server this WebAppContext can see -->
  <Get name="serverClassMatcher">
    <Call name="add">
      <!-- needed for all example servlets to actually see LowResourceMonitor -->
      <Arg>-org.eclipse.jetty.server.LowResourceMonitor</Arg>
    </Call>
    <Call name="add">
      <!-- needed for example servlets that use Server.getBean().
           FromBaseRequestServlet
           and FromRequestAttributeChannelServlet
           and FromRequestAttributeConnectionServlet -->
      <Arg>-org.eclipse.jetty.server.Server</Arg>
    </Call>
    <Call name="add">
      <!-- needed for FromBaseRequestServlet -->
      <Arg>-org.eclipse.jetty.server.Request</Arg>
    </Call>
  </Get>
</Configure>
```

These changes are not applied to the WebAppClassloader.
To work, the project needs the following change.

``` diff
diff --git a/standalone/lowresource-access/webapps/demo.xml b/standalone/lowresource-access/webapps/demo.xml
index c5cf4dc..4752183 100644
--- a/standalone/lowresource-access/webapps/demo.xml
+++ b/standalone/lowresource-access/webapps/demo.xml
@@ -20,7 +20,7 @@
     </Arg>
   </Call>
   <!-- tweak what classes from the Server this WebAppContext can see -->
-  <Get name="serverClassMatcher">
+  <Get name="hiddenClassMatcher">
     <Call name="add">
       <!-- needed for all example servlets to actually see LowResourceMonitor -->
       <Arg>-org.eclipse.jetty.server.LowResourceMonitor</Arg>
```

The project at https://github.com/jetty/jetty-examples/tree/12.0.x/standalone/lowresource-access can be used as an example.
Note: you will need to revert the commit https://github.com/jetty/jetty-examples/commit/14d5c894adaa2d68066147434bd1a2c31a6f3756 to replicate this bug.

### PR DESCRIPTION
**Pull request by gregw**
Fix #11767 by making deprecated ClassMatcher class wrap the util ClassMatcher

### CODE REVIEW

**Comment by gregw** [comment]
The constructor function is a little clumsy, but perhaps better than exposing the internals.

**Review by joakime — DISMISSED** [review]
Looks good.  I also added some missing ee9 tests which are showing as green too.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ClassMatcher.java [hunk 1]
```diff
@@ -525,6 +525,24 @@ public ClassMatcher(String pattern)
         add(pattern);
     }
 
+    protected interface Constructor<T extends ClassMatcher>
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ClassMatcher.java:529]: I would deprecate already this interface.
When we remove the `eeN` versions of `ClassMatcher`, we want to remove this interface also, as it won't be necessary anymore.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ClassMatcher.java [hunk 2]
```diff
@@ -525,6 +525,24 @@ public ClassMatcher(String pattern)
...
+
+    /**
+     * Wrap an instance of a {@link ClassMatcher} using a constructor of an extended {@code ClassMatcher}
+     * that needs access to the internal fields of the passed matcher.
+     * @param matcher The matcher to wrap
+     * @param constructor The constructor to build the API specific wrapper
+     * @param <T> The type of the API specific wrapper
+     * @return A wrapper of the {@code matcher}, sharing internal state.
+     */
+    protected static <T extends ClassMatcher> T wrap(ClassMatcher matcher, Constructor<T> constructor)
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ClassMatcher.java:544]: Ditto re deprecation.

File: jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/ClassMatcher.java [hunk 3]
```diff
@@ -61,4 +61,16 @@ public ClassMatcher asImmutable()
...
             _locations.asImmutable());
     }
+
+    /**
+     * Create a {@link ClassMatcher webapp ClassMatcher} that wraps a {@link org.eclipse.jetty.util.ClassMatcher util ClassMatcher}
+     * for deprecated API usage.
+     * @param matcher The util {@link org.eclipse.jetty.util.ClassMatcher} to wrap
+     * @return A {@link ClassMatcher webapp ClassMatcher}
+     */
+    public static ClassMatcher wrap(org.eclipse.jetty.util.ClassMatcher matcher)
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/ClassMatcher.java]: Can be package-private (just remove `public`).

**Review by sbordet — CHANGES_REQUESTED** [review]

**Review by sbordet — APPROVED** [review]

**Pull request merged**